atlasdb-autobatch/src/main/java/com/palantir/atlasdb/autobatch/Autobatchers.java

```diff
@@ -0,0 +1,100 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.autobatch;
+
+import java.util.List;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+import javax.annotation.Nullable;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import com.lmax.disruptor.EventHandler;
+
+public final class Autobatchers {
+
+    private static final int DEFAULT_BUFFER_SIZE = 1024;
+
+    private Autobatchers() {}
+
+    /**
+     * This coalesces requests for the same input to the same output. {@link #independent} differs from this method
+     * because the same semantic request will be processed independently.
+     *
+     * @param function function that processes a batch ({@link java.util.Set}) of {@code I}, producing a mapping from
+     * {@code I} to {@code O}
+     * @param <I> type of input element
+     * @param <O> type of output element
+     * @return builder where the autobatcher can be further customised
+     * @see CoalescingRequestFunction
+     * @see CoalescingRequestConsumer
+     */
+    public static <I, O> AutobatcherBuilder<I, O> coalescing(CoalescingRequestFunction<I, O> function) {
+        return new AutobatcherBuilder<>(bufferSize -> new CoalescingBatchingEventHandler<>(function, bufferSize));
+    }
+
+    /**
+     * This builds an autobatcher where the function takes a batch of elements (equal elements are considered
+     * independently) and completes the futures associated with each request. This is different from {@link #coalescing}
+     * as each input element even if identical, is processed independently and separately.
+     *
+     * @param batchFunction consumer that processes a batch of input elements and resolves the associated futures with
+     * the result for that input element
+     * @param <I> type of input element
+     * @param <O> type of output element
+     * @return builder where the autobatch can be further customised
+     */
+    public static <I, O> AutobatcherBuilder<I, O> independent(Consumer<List<BatchElement<I, O>>> batchFunction) {
+        return new AutobatcherBuilder<>(bufferSize -> new IndependentBatchingEventHandler<>(batchFunction, bufferSize));
+    }
+
+    public static final class AutobatcherBuilder<I, O> {
+
+        private final Function<Integer, EventHandler<BatchElement<I, O>>> handlerFactory;
+
+        @Nullable private String purpose;
+        private int bufferSize = DEFAULT_BUFFER_SIZE;
+
+        private AutobatcherBuilder(Function<Integer, EventHandler<BatchElement<I, O>>> handlerFactory) {
+            this.handlerFactory = handlerFactory;
+        }
+
+        public AutobatcherBuilder<I, O> safeLoggablePurpose(String purposeParam) {
+            this.purpose = purposeParam;
+            return this;
+        }
+
+        @VisibleForTesting
+        AutobatcherBuilder<I, O> bufferSize(int bufferSizeParam) {
```

Comment:
It doesn't look like this is set anywhere - I guess you probably need it for a future test, but should check that this isn't a leftover artifact

Reply:
might remove this, test of function didn't need it in the end